### ISSUE TITLE
Missing an equivalent to cudaDeviceSynchronize

### ISSUE THREAD

**Issue by TApplencourt** [KhronosGroup/SYCL-Docs#284]
In our quest to find missing CUDA API equivalents in SYCL we found a new one: [cudaDeviceSynchronize](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html#group__CUDART__DEVICE_1g10e20b05a95f638a4071a655503df25d)

This function is pretty common. [Github returns 308k matches for the API](https://github.com/search?q=cudaDeviceSynchronize&type=code).

Of course, the "correct" pattern in SYCL is to pass a queue and use it to sync; Not to query for the device over and over again and wait on it. But we may want to ease the porting from CUDA. Refactoring your code to use queues may be icky. So I propose to add a  `wait()` and friend in the `device`.  I propose us to be more descriptive than prescriptive :)

TLDR:
CUDA has `cudaDeviceSynchronize`, we have no equivalent. I think we should. What do people think?

**Comment by tomdeakin** [comment]
Don’t we always have a queue though… How is anything submitted to the device without one?

The `cudaDeviceSynchronize()` function only synchronises across all streams for the *current* device, so with multiple devices you need to query the device in CUDA too with `cudaSetDevice(int)`.

In SYCL the equivalent use is a wait on **all** queues to the same device. I could see an argument for some shortcut function which is a wait on all queues set up for a particular device. But I don’t know if that’s the use case in applications… are they really using multiple queues?

**Comment by TApplencourt** [comment]
>Don’t we always have a queue though… How is anything submitted to the device without one?

You are ofc, right. To migrate a code like this
```
foo0() {
 <<<kernel>>>
}
foo1() {
cudaDeviceSynchronize()
}
main() {
 foo0();
 foo1();
}
```
One needs to either change the function signature to take a queue (nobody does that it's too painful in a large codebase) or to create a queue singleton (not the hardest thing in the world, but not the easiest either. Need to be careful about race condition and the like)

With device Q, one should able to do:
```
foo0() {
sycl::Queue Q; # Creating a queue over and over again, but that ok. If we have "default context" should not be too expensive.
Q.submit
}
foo1() {
sycl::device D;
D.wait()
}
```
The porting is "trivial".

>But I don’t know if that’s the use case in applications… are they really using multiple queues?

Some application does. Cuda streams are in-order, so if your kernel isn't big enough to saturate the device,  you can use multiple streams to run kernels concurrently.

**Comment by tomdeakin** [comment]
It's a bit nasty, but if you wanted to replicate the CUDA behaviour you could do a `static sycl::queue Q{sycl::property_list{sycl::property::queue::in_order{}}}` somewhere in the program. This would solve the first pattern you gave, and mean you only need one queue, and can synchronise on it anywhere.
```
static sycl::queue Q{sycl::property_list{sycl::property::queue::in_order{}}};

foo0() {
Q.submit(...);
}

foo1() {
Q.wait();
}
```

I guess we're lucky the device default constructor and the queue default constructor use the save default device selector, and that should deterministically select the same device.

We'd need to ask the implementors if they track the queues for each device - that info would be needed to do a device::wait() routine.

**Comment by TApplencourt** [comment]
>It's a bit nasty, but if you wanted to replicate the CUDA behaviour you could do a static sycl::queue Q{sycl::property_list{sycl::property::queue::in_order{}}} somewhere in the program. This would solve the first pattern you gave, and mean you only need one queue, and can synchronise on it anywhere.

Yes, exactly. In SYCLoMatic they call use [get_default_queue](https://github.com/oneapi-src/SYCLomatic/blob/e8a309cbda90ab10117bce657879267966fe6d5e/clang/test/dpct/helper_files_ref/include/device.hpp#L479). 
A little more complicated than static, but this is the way to do it right now indeed! 

`device.wait()` will remove the need for this singleton. 

>We'd need to ask the implementors if they track the queues for each device - that info would be needed to do a device::wait() routine.

Agree! I know the cuda backend for LLVM does, but I guess it's not super important  for this particular backend :p

**Comment by TApplencourt** [comment]
After more thinking (thanks @Kerilk), maybe we want more a `context.wait()`, who will wait on all devices of the context to finish. I look more to the cuda drivers API [cuCtxSynchronize](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__CTX.html#group__CUDA__CTX_1g7a54725f28d34b8c6299f0c6ca579616)

```
foo0() {
sycl::Queue Q; # Creating a queue over and over again, but that ok. If we have "default context," should not be too expensive.
Q.submit
}
foo1() {
sycl::context C; # get the default context used by the default queue;
C.wait()
}
```
